plugins/api/src/atomic_apply_metadata_operations.cpp

```diff
@@ -0,0 +1,590 @@
+            throw std::runtime_error{fmt::format("Invalid entity type specified [entity_type => {}]", _entity_type)};
+        }
+
+        for (auto&& row : irods::query{&_comm, gql}) {
+            return std::stoi(row[0]);
+        }
+
+        throw std::runtime_error{fmt::format("Entity does not exist [entity_name => {}]", _entity_name)};
+    }
+
+    auto new_database_connection() -> nanodbc::connection
+    {
+        const std::string dsn = [] {
+            if (const char* dsn = std::getenv("irodsOdbcDSN"); dsn) {
+                return dsn;
+            }
+
+            return "iRODS Catalog";
+        }();
+
+        std::string config_path;
+
+        if (const auto error = irods::get_full_path_for_config_file("server_config.json", config_path); !error.ok()) {
+            log::api::error("Server configuration not found");
+            throw std::runtime_error{"Failed to connect to catalog"};
+        }
+
+        log::api::trace("Reading server configuration ...");
+
+        json config;
+
+        {
+            std::ifstream config_file{config_path};
+            config_file >> config;
+        }
+
+        try {
+            const auto& db_plugin_config = config.at(irods::CFG_PLUGIN_CONFIGURATION_KW).at(irods::PLUGIN_TYPE_DATABASE);
+            const auto& db_instance = *std::begin(db_plugin_config);
+            const auto db_username = db_instance.at(irods::CFG_DB_USERNAME_KW).get<std::string>();
+            const auto db_password = db_instance.at(irods::CFG_DB_PASSWORD_KW).get<std::string>();
+
+            return {dsn, db_username, db_password};
+        }
+        catch (const std::exception& e) {
+            log::api::error(e.what());
+            throw std::runtime_error{"Failed to connect to catalog"};
+        }
+    }
+
+    auto user_has_permission_to_modify_metadata(rsComm_t& _comm,
+                                                nanodbc::connection& _db_conn,
+                                                int _object_id,
+                                                const std::string& _entity_type) -> bool
+    {
+        if (_entity_type == "data_object" || _entity_type == "collection") {
+            const auto query = fmt::format("select t.token_id from R_TOKN_MAIN t"
+                                           " inner join R_OBJT_ACCESS a on t.token_id = a.access_type_id "
+                                           "where"
+                                           " a.user_id = (select user_id from R_USER_MAIN where user_name = '{}') and"
+                                           " a.object_id = '{}'", _comm.clientUser.userName, _object_id);
+
+            if (auto row = execute(_db_conn, query); row.next()) {
+                constexpr int access_modify_object = 1120;
+                return row.get<int>(0) >= access_modify_object;
+            }
+        }
+        else if (_entity_type == "user" || _entity_type == "resource") {
+            return irods::is_privileged_client(_comm);
```

Comment:
Should applying metadata to users and resources require admin-level privileges? Can rodsusers apply metadata to users and resources now?

Reply:
Based on `imeta`'s docs, the answer is users must be admins to set metadata on users and resources.

Reply:
yes, confirmed.